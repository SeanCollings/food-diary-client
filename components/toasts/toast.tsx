import {
  TToastStatus,
  ToastStatus,
  ToastWithId,
  useToast,
} from '@store/toast-context';
import { FC, useEffect, useRef } from 'react';
import styled from 'styled-components';
import { MdErrorOutline, MdOutlineCheck, MdClose } from 'react-icons/md';
import { getClassNames } from '@utils/string-utils';
import { APP_THEME_DEFAULT, MEDIA_MOBILE } from '@utils/app.constants';

const TOAST_HEIGHT = 78;
const MAX_WIDTH = 416;
const TOP_PADDING = 4;
export const HIDE_TOAST_TIMER = 5000;

interface ISToastContainer {
  animateStyle: AnimateStyle;
  toastPosition: number;
}

const SToastContainer = styled.div<ISToastContainer>`
  position: fixed;
  min-height: ${TOAST_HEIGHT}px;
  display: table;
  width: 90%;
  max-width: ${MAX_WIDTH}px;
  box-shadow: 0px 4px 8px rgba(35, 35, 35, 0.2);
  border-radius: 4px;
  color: ${APP_THEME_DEFAULT.textLight};

  ${({ toastPosition }) => `top: ${toastPosition + TOP_PADDING}px`};
  z-index: 1000;

  animation-name: ${({ animateStyle }) => animateStyle.type};
  animation-duration: 0.1s;
  animation-timing-function: ease-out;
  animation-fill-mode: forwards;

  right: 2rem;
  ${MEDIA_MOBILE} {
    right: 1rem;
  }

  @keyframes enter {
    from {
      opacity: 0;
      right: -${MAX_WIDTH / 4}px;
    }
    to {
      opacity: 1;
    }
  }

  @keyframes exit {
    from {
      opacity: 1;
    }
    to {
      opacity: 0;
      right: -${MAX_WIDTH / 4}px;
    }
  }
`;
const SInnerContainer = styled.div`
  display: flex;
  flex-direction: row;
  border-radius: 4px;
  padding: 16px;

  &.success {
    background-color: var(--success);
  }
  &.error {
    background-color: var(--error);
  }
`;
const SContentWrapper = styled.div`
  flex: 1;
  padding-left: 8px;
  padding-right: 8px;
`;
const SHeader = styled.div`
  display: flex;
  align-items: center;
`;
const SHeaderContainer = styled(SHeader)`
  justify-content: space-between;
`;
const SHeaderH5 = styled.h5`
  margin: 0;
  font-size: 16px;
  line-height: 1.8;
`;
const SMessage = styled.p`
  margin: 0;
  font-size: 14px;
`;
const SMdClose = styled(MdClose)`
  cursor: pointer;
  :hover {
    opacity: 0.6;
  }
`;

type AnimateStyleType = 'enter' | 'exit';

interface AnimateStyle {
  type: AnimateStyleType;
}

interface ToastProps {
  toastId: string;
  toast: ToastWithId;
  toastPosition: number;
}

const getToastIcon = (type: TToastStatus) =>
  ({
    [ToastStatus.SUCCESS]: MdOutlineCheck,
    [ToastStatus.ERROR]: MdErrorOutline,
  }[type]);

export const Toast: FC<ToastProps> = ({ toast, toastId, toastPosition }) => {
  const toastRef = useRef<HTMLDivElement>(null);
  const { removeToast, hideToast } = useToast();

  useEffect(() => {
    let timer: ReturnType<typeof setTimeout>;
    const toastRefCurrent = toastRef.current;

    const animationHasEnded = () => {
      if (toast.isHiding) {
        removeToast(toastId);
      }
    };

    if (toast.timeDismissed) {
      timer = setTimeout(() => {
        hideToast(toastId);
      }, HIDE_TOAST_TIMER);
    }

    toastRefCurrent?.addEventListener('animationend', () =>
      animationHasEnded(),
    );

    return () => {
      clearTimeout(timer);
      toastRefCurrent?.addEventListener('animationend', animationHasEnded);
    };
  }, [hideToast, removeToast, toast.isHiding, toast.timeDismissed, toastId]);

  const animateStyle: AnimateStyle = toast.isHiding
    ? { type: 'exit' }
    : { type: 'enter' };

  const ToastIcon = getToastIcon(toast.status);
  const classColour = getClassNames({
    success: toast.status === 'success',
    error: toast.status === 'error',
  });

  return (
    <SToastContainer
      ref={toastRef}
      animateStyle={animateStyle}
      toastPosition={toastPosition}
      id={`toast-${toastId}`}
      data-testid={`toast-${toastId}`}
    >
      <SInnerContainer className={classColour}>
        <ToastIcon size={28} color={APP_THEME_DEFAULT.textLight} />
        <SContentWrapper>
          <SHeaderContainer>
            <SHeader>
              <SHeaderH5>{toast.title}</SHeaderH5>
            </SHeader>
          </SHeaderContainer>
          <SMessage>{toast.message}</SMessage>
        </SContentWrapper>
        <SMdClose
          size={28}
          color={APP_THEME_DEFAULT.textLight}
          onClick={() => hideToast(toast.id)}
          data-testid="toast-dismiss"
        />
      </SInnerContainer>
    </SToastContainer>
  );
};
