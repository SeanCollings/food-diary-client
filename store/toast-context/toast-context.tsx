import {
  FC,
  ReactNode,
  createContext,
  useCallback,
  useContext,
  useMemo,
  useState,
} from 'react';
import { getUniqueId } from '@utils/unique-id';

export enum ToastStatus {
  SUCCESS = 'success',
  ERROR = 'error',
}

export type TToastStatus = `${ToastStatus}`;

export interface Toast {
  title: string;
  message: string;
  status: TToastStatus;
  isHiding?: boolean;
  timeDismissed?: boolean;
}

export interface ToastWithId extends Toast {
  id: string;
}

export type ToastData = {
  [id in string]: ToastWithId;
};

interface ToastContextProps {
  children: ReactNode;
}

type TToastContext = ReturnType<typeof useToastContext>;

const ToastContext = createContext<TToastContext>({} as TToastContext);

const useToastContext = () => {
  const [activeToasts, setActiveToasts] = useState<ToastData>({});

  const showToast = useCallback((notificatioData: Toast) => {
    const id = getUniqueId();

    setActiveToasts((currentToasts) => {
      const updatedToast = {
        ...currentToasts,
        [id]: { ...notificatioData, id },
      };
      return updatedToast;
    });
    return id;
  }, []);

  const removeToast = useCallback(
    (id: string) => {
      setActiveToasts((currentToasts) => {
        const updatedToast = { ...currentToasts };
        delete updatedToast[id];
        return updatedToast;
      });
    },
    [setActiveToasts],
  );

  const hideToast = useCallback(
    (id: string) => {
      setActiveToasts((currentToasts) => {
        const updatedToast = { ...currentToasts };
        updatedToast[id] = {
          ...updatedToast[id],
          isHiding: true,
        };
        return updatedToast;
      });
    },
    [setActiveToasts],
  );

  const context = useMemo(
    () => ({
      toasts: activeToasts,
      showToast,
      removeToast,
      hideToast,
    }),
    [activeToasts, showToast, removeToast, hideToast],
  );

  return context;
};

export const ToastContextProvider: FC<ToastContextProps> = ({ children }) => {
  return (
    <ToastContext.Provider value={useToastContext()}>
      {children}
    </ToastContext.Provider>
  );
};

export const useToast = () => useContext<TToastContext>(ToastContext);
