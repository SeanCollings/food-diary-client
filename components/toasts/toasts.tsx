import { ToastData, useToast } from '@store/toast-context';
import { FC, memo, useEffect, useState } from 'react';
import { Toast } from '@components/toasts';

const TOAST_MARGIN_BOTTOM = 16;
const TOTAL_TOASTS_AT_ONCE = 5;

interface ToastsProps {
  toasts: ToastData;
}

export const Toasts: FC<ToastsProps> = memo(({ toasts }) => {
  const [allToastHeights, setAllToastHeights] = useState<number[]>([]);

  const allToastIds = Object.keys(toasts);

  useEffect(() => {
    if (allToastIds.length !== allToastHeights.length) {
      const allHeights = allToastIds.map((toastId) => {
        const toastElement = document.getElementById(`toast-${toastId}`);

        if (toastElement) {
          return toastElement.offsetHeight;
        }

        return 0;
      });

      setAllToastHeights(allHeights);
    }
  }, [allToastHeights.length, allToastIds]);

  const getPreviousToastHeights = (currentIndex: number) =>
    allToastHeights.reduce((totalHeight, height, index) => {
      if (index < currentIndex) {
        totalHeight += height;
      }

      return totalHeight;
    }, 0);

  const getTotalMarginTop = (index: number) =>
    (index + 1) * TOAST_MARGIN_BOTTOM;

  const getToastPosition = (index: number) =>
    getPreviousToastHeights(index) + getTotalMarginTop(index);

  if (!allToastIds.length) {
    return null;
  }

  return (
    <div>
      {allToastIds.map((toastId, index) => {
        const toast = toasts[toastId];

        if (index + 1 > TOTAL_TOASTS_AT_ONCE) {
          return null;
        }

        return (
          <Toast
            key={toastId}
            toastId={toastId}
            toast={toast}
            toastPosition={getToastPosition(index)}
          />
        );
      })}
    </div>
  );
});
Toasts.displayName = 'Toasts';

export const DisplayToasts: FC = () => {
  const { toasts } = useToast();

  if (!Object.keys(toasts).length) {
    return null;
  }

  return <Toasts toasts={toasts} />;
};
