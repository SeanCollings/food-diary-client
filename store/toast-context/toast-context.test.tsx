import { render } from '@testing-library/react';
import { ToastContextProvider, useToast } from './toast-context';
import { useEffect, useRef } from 'react';
import { Toasts } from '@components/toasts';

jest.useFakeTimers();
jest.mock('@utils/unique-id', () => ({
  getUniqueId: () => 'randomId',
}));

const TestComponent = () => <div>Test component</div>;

const MockComponentShowToast = () => {
  const stateUpdateRef = useRef<number>(0);
  const { toasts, showToast } = useToast();

  useEffect(() => {
    stateUpdateRef.current += 1;

    if (!Object.keys(toasts).length) {
      const id = showToast({
        message: 'Mock message.',
        title: 'Mock Title',
        status: 'success',
      });
      expect(id).toBeDefined();
    }

    if (stateUpdateRef.current === 2) {
      expect(toasts).toMatchInlineSnapshot(`
        {
          "randomId": {
            "id": "randomId",
            "message": "Mock message.",
            "status": "success",
            "title": "Mock Title",
          },
        }
      `);
    }
  }, [toasts, showToast]);

  return <Toasts toasts={toasts} />;
};

const MockComponentHideToast = () => {
  const firstRenderRef = useRef<boolean>(false);
  const { toasts, showToast, hideToast } = useToast();

  useEffect(() => {
    if (!firstRenderRef.current) {
      if (!Object.keys(toasts).length) {
        showToast({
          message: 'Mock message.',
          title: 'mock Title',
          status: 'success',
        });
      } else {
        firstRenderRef.current = true;
        hideToast('randomId');
      }
    } else if (toasts['randomId']?.isHiding) {
      expect(toasts['randomId'].isHiding).toBeTruthy();
    }
  }, [toasts, showToast, hideToast]);

  return <Toasts toasts={toasts} />;
};

const MockComponentRemoveToast = () => {
  const firstRenderRef = useRef<boolean>(false);
  const { toasts, showToast, removeToast } = useToast();

  useEffect(() => {
    if (!firstRenderRef.current) {
      if (!toasts) {
        showToast({
          message: 'Mock message.',
          title: 'Mock Title',
          status: 'success',
        });
      } else {
        firstRenderRef.current = true;
        removeToast('test');
      }
    }
  }, [toasts, showToast, removeToast]);

  return <Toasts toasts={toasts} />;
};

describe('store - toast-context', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render provider with children', () => {
    const { asFragment } = render(
      <ToastContextProvider>
        <TestComponent />
      </ToastContextProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should show Toast', () => {
    const { asFragment } = render(
      <ToastContextProvider>
        <MockComponentShowToast />
      </ToastContextProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should hide Toast', () => {
    const { asFragment } = render(
      <ToastContextProvider>
        <MockComponentHideToast />
      </ToastContextProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should remove Toast', () => {
    const { asFragment } = render(
      <ToastContextProvider>
        <MockComponentRemoveToast />
      </ToastContextProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
