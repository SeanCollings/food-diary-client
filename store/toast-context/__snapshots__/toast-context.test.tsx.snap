// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store - toast-context should hide Toast 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-c2b481e2-0 dGESEi"
      data-testid="toast-randomId"
      id="toast-randomId"
    >
      <div
        class="sc-c2b481e2-1 haSkKn success"
      >
        <svg
          color="#ffffff"
          fill="currentColor"
          height="28"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(255, 255, 255);"
          viewBox="0 0 24 24"
          width="28"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41L9 16.17z"
          />
        </svg>
        <div
          class="sc-c2b481e2-2 cOpAYz"
        >
          <div
            class="sc-c2b481e2-3 sc-c2b481e2-4 lhRooE kxINyd"
          >
            <div
              class="sc-c2b481e2-3 lhRooE"
            >
              <h5
                class="sc-c2b481e2-5 dlbyEI"
              >
                mock Title
              </h5>
            </div>
          </div>
          <p
            class="sc-c2b481e2-6 estfx"
          >
            Mock message.
          </p>
        </div>
        <svg
          class="sc-c2b481e2-7 qfhLu"
          color="#ffffff"
          data-testid="toast-dismiss"
          fill="currentColor"
          height="28"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(255, 255, 255);"
          viewBox="0 0 24 24"
          width="28"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`store - toast-context should remove Toast 1`] = `<DocumentFragment />`;

exports[`store - toast-context should render provider with children 1`] = `
<DocumentFragment>
  <div>
    Test component
  </div>
</DocumentFragment>
`;

exports[`store - toast-context should show Toast 2`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-c2b481e2-0 cjZdCO"
      data-testid="toast-randomId"
      id="toast-randomId"
    >
      <div
        class="sc-c2b481e2-1 haSkKn success"
      >
        <svg
          color="#ffffff"
          fill="currentColor"
          height="28"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(255, 255, 255);"
          viewBox="0 0 24 24"
          width="28"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41L9 16.17z"
          />
        </svg>
        <div
          class="sc-c2b481e2-2 cOpAYz"
        >
          <div
            class="sc-c2b481e2-3 sc-c2b481e2-4 lhRooE kxINyd"
          >
            <div
              class="sc-c2b481e2-3 lhRooE"
            >
              <h5
                class="sc-c2b481e2-5 dlbyEI"
              >
                Mock Title
              </h5>
            </div>
          </div>
          <p
            class="sc-c2b481e2-6 estfx"
          >
            Mock message.
          </p>
        </div>
        <svg
          class="sc-c2b481e2-7 qfhLu"
          color="#ffffff"
          data-testid="toast-dismiss"
          fill="currentColor"
          height="28"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(255, 255, 255);"
          viewBox="0 0 24 24"
          width="28"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
